sdcflows/fieldmaps.py

```diff
@@ -0,0 +1,345 @@
+    ...     testdata_dir / "sub-01" / "fmap" / "sub-01_dir-LR_epi.nii.gz",
+    ... )
+    >>> f.metadata['TotalReadoutTime']
+    0.005
+
+    >>> f = FieldmapFile(
+    ...     testdata_dir / "sub-01" / "fmap" / "sub-01_dir-LR_epi.nii.gz",
+    ...     metadata={'TotalReadoutTime': 0.006}
+    ... )
+    >>> f.metadata['TotalReadoutTime']
+    0.006
+
+    >>> FieldmapFile(
+    ...     testdata_dir / "sub-01" / "fmap" / "sub-01_phasediff.nii.gz",
+    ...     find_meta=False
+    ... )  # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+    MetadataError:
+
+    >>> f = FieldmapFile(
+    ...     testdata_dir / "sub-01" / "fmap" / "sub-01_phasediff.nii.gz"
+    ... )
+    >>> f.metadata['EchoTime2']
+    0.00746
+
+    >>> FieldmapFile(
+    ...     testdata_dir / "sub-01" / "fmap" / "sub-01_phase2.nii.gz",
+    ...     find_meta=False
+    ... )  # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+    MetadataError:
+
+    >>> f = FieldmapFile(
+    ...     testdata_dir / "sub-01" / "fmap" / "sub-01_phase2.nii.gz"
+    ... )
+    >>> f.metadata['EchoTime']
+    0.00746
+
+    >>> FieldmapFile(
+    ...     testdata_dir / "sub-01" / "fmap" / "sub-01_fieldmap.nii.gz",
+    ...     find_meta=False
+    ... )  # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+    MetadataError:
+
+    >>> f = FieldmapFile(
+    ...     testdata_dir / "sub-01" / "fmap" / "sub-01_fieldmap.nii.gz"
+    ... )
+    >>> f.metadata['Units']
+    'rad/s'
+
+    """
+
+    path = attr.ib(converter=Path, repr=str, on_setattr=attr.setters.NO_OP)
+    """Path to a fieldmap file."""
+
+    entities = attr.ib(init=False, repr=False)
+    """BIDS entities extracted from filepath."""
+
+    suffix = attr.ib(init=False, repr=False)
+    """Extracted suffix from input file."""
+
+    bids_root = attr.ib(init=False, default=None, repr=False)
+    """Path of the BIDS root."""
+
+    metadata = attr.ib(kw_only=True, default=attr.Factory(dict))
+    """
+    Metadata associated to this file. When provided as keyword argument in initialization,
+    will overwrite metadata read from sidecar JSON.
+    """
+
+    find_meta = attr.ib(kw_only=True, default=True, type=bool, repr=False)
+    """Enable/disable automated search for corresponding metadata."""
+
+    @path.validator
+    def check_path(self, attribute, value):
+        """Validate a fieldmap path."""
+        if isinstance(value, BIDSFile):
+            value = Path(value.path)
+        if isinstance(value, str):
+            value = Path(value)
+
+        if not value.is_file():
+            raise FileNotFoundError(
+                f"File path <{value}> does not exist, is a broken link, or it is not a file"
+            )
+
+        if not str(value).endswith((".nii", ".nii.gz")):
+            raise ValueError(f"File path <{value}> does not look like a NIfTI file.")
+
+        suffix = str(value).rpartition(".nii")[0].rpartition("_")[-1]
```

Comment:
will require adding an import, but I think this would be cleaner as a regex
```suggestion
        suffix = re.search('(?<=_)\w+(?=.nii)', str(value)).group()
```

Reply:
I'm always wary of regexes, if we're to use them then we need to test them

Reply:
I tested with all current modality keys with unix / windows paths and it extracted fine - any other tests you can think of?
and just to polish this up a little
```suggestion
        suffix = re.search('(?<=_)\w+(?=.nii)', str(value.name)).group()
```

Reply:
The fact that I had to modify a little your suggestion does not help my tranquility :D : https://github.com/nipreps/sdcflows/blob/f551841b92605b5475e898b385c2774afcb41490/sdcflows/fieldmaps.py#L175